chainerx_cc/examples/mnist_cc/train_mnist.cc

```diff
@@ -0,0 +1,203 @@
+                p.ClearGrad();
+            }
+        }
+
+        // Evaluate.
+        {
+            chainerx::NoBackpropModeScope scope{};
+
+            chainerx::Array loss = chainerx::Zeros({}, chainerx::Dtype::kFloat32);
+            chainerx::Array acc = chainerx::Zeros({}, chainerx::Dtype::kFloat32);
+
+            for (int64_t i = 0; i < n_test; i += batch_size) {
+                std::vector<chainerx::ArrayIndex> indices{chainerx::Slice{i, i + batch_size}};
+                chainerx::Array x = test_x.At(indices);
+                chainerx::Array t = test_t.At(indices);
+                chainerx::Array y = model(x);
+                loss += SoftmaxCrossEntropy(y, t, false);
+                acc += (y.ArgMax(1).AsType(t.dtype()) == t).Sum().AsType(acc.dtype());
+            }
+
+            std::cout << "epoch: " << epoch << " loss=" << chainerx::AsScalar(loss / n_test)
+                      << " accuracy=" << chainerx::AsScalar(acc / n_test)
+                      << " elapsed_time=" << std::chrono::duration<double>{std::chrono::high_resolution_clock::now() - start}.count()
+                      << std::endl;
+        }
+    }
+}
+
+int main(int argc, char** argv) {
+    int64_t epochs{20};
+    int64_t batch_size{100};
+    int64_t n_hidden{1000};
+    int64_t n_layers{3};
+    float lr{0.01};
+    std::string device_name{"native"};
+    std::string mnist_root{"./"};
+
+    for (int i = 1; i < argc; i += 2) {
+        std::string arg = argv[i];
+        if (arg == "--epoch") {
+            epochs = std::atoi(argv[i + 1]);
+        } else if (arg == "--batchsize") {
+            batch_size = std::atoi(argv[i + 1]);
+        } else if (arg == "--n-hidden") {
+            n_hidden = std::atoi(argv[i + 1]);
+        } else if (arg == "--n-layers") {
+            n_layers = std::atoi(argv[i + 1]);
+        } else if (arg == "--lr") {
+            lr = std::atof(argv[i + 1]);
+        } else if (arg == "--device") {
+            device_name = argv[i + 1];
+        } else if (arg == "--data") {
+            mnist_root = argv[i + 1];
+        } else {
+            throw std::runtime_error("Unknown argument: " + arg);
+        }
+    }
+
+    if (mnist_root.back() != '/') {
+        mnist_root += "/";
+    }
+
+    chainerx::Context ctx{};
+    chainerx::SetDefaultContext(&ctx);
+    chainerx::Device& device = ctx.GetDevice(device_name);
+    chainerx::SetDefaultDevice(&device);
+
+    std::cout << "Epochs: " << epochs << std::endl;
+    std::cout << "Minibatch size: " << batch_size << std::endl;
+    std::cout << "Hidden neurons: " << n_hidden << std::endl;
+    std::cout << "Layers: " << n_layers << std::endl;
+    std::cout << "Learning rate: " << lr << std::endl;
+    std::cout << "Device: " << device.name() << std::endl;
+    std::cout << "MNIST root: " << mnist_root << std::endl;
+
+    RunWithDefaultDevice(epochs, batch_size, n_hidden, n_layers, lr, mnist_root);
```

Comment:
I think `WithDefaultDevice` part is nonessential. Simply `Run` seems better for me.

Reply:
Fixed.